SD-K9/src/app/providers/firebase/google-calendar.service.ts

```diff
@@ -0,0 +1,123 @@
+  // Permission: access type to google calendar
+  private _SCOPES = "https://www.googleapis.com/auth/calendar";
+
+
+  constructor(
+    private _angularFireAuth: AngularFireAuth,
+    private _http: HttpClient,
+    private _calendarListStore: CalendarListStore
+    // private _angularFireStore: AngularFirestore // TODO: To remove in this service, but currently serves as a reminder that we have FireStore
+  ) {
+    this._initClient();
+    this.user$ = this._angularFireAuth.authState;
+  }
+
+  private _initClient() {
+    gapi.load('client:auth2', () => {
+      console.log('loaded client');
+
+      gapi.client.init({
+        apiKey: this._API_KEY,
+        clientId: this._CLIENT_ID,
+        discoveryDocs: this._DISCOVERY_DOCS,
+        scope: this._SCOPES
+      });
+
+      // Important to let gapi.client know user is authorized
+      gapi.auth2.init({
+        clientId: this._CLIENT_ID,
+        scope: this._SCOPES
+      })
+
+      gapi.client.load('calendar', 'v3', () => console.log('loaded calendar client'));
+    });
+  }
+
+  async googleSignin() {
+    const googleAuth = gapi.auth2.getAuthInstance();
+    const googleUser = await googleAuth.signIn().catch((error) => {console.log(error)});
+
+    this._authResponse = googleUser.getAuthResponse();
+
+    const token = googleUser.getAuthResponse().id_token;
+    const credential = firebase.auth.GoogleAuthProvider.credential(token);
+    await this._angularFireAuth.signInWithCredential(credential);
+  }
+
+  async signOut() {
+    await this._angularFireAuth.signOut();
+    this._calendarListStore.clear();
+  }
+
+  async getCalendarList() {
+    // clear calendar list
+    this._calendarListStore.clear();
+
+    let calendarIds: string[] = [];
+
+    // while google auth2 does not publicly provide a refresh token, request user to sign in.
+    if (!this._authResponse) {
+      await this.googleSignin();
+    }
+
+    // while google calendar api does not provide a special keyword to get all calendarIds
+    // request to get a list of calendars via HttpClient
+    let reqHeaders = new HttpHeaders({
+      'Content-Type': 'application/json; charset=utf-8',
+      'cache-control': 'private',
+      'Authorization': `${this._authResponse.token_type} ${this._authResponse.access_token}`
+    })
+    await this._http.get(`https://www.googleapis.com/calendar/v3/users/me/calendarList`, {headers: reqHeaders})
+    .subscribe(resource => {   
+      let items = resource['items'];
+
+      items.forEach(async(item) => {
+        let calendarId: string = item['id'];
+
+        this._calendarListStore.storeCalendarId({id: calendarId, checked: true});
+        calendarIds.push(calendarId);
+      });
+    });
+
+    return calendarIds;
+  }
+
+  async getEvents(calendarId: string) {
+    const events = await gapi.client.calendar.events.list({
+      calendarId: calendarId,
+      timeMin: (new Date()).toISOString(),
+      showDeleted: false,
+      singleEvents: true,
+      // maxResults: 10,
```

Comment:
Does this mean the app can only gather 10 events at a time from the calendar?

Reply:
No, it is commented-out.
It's useful to uncomment when you want to do a quick test to validate the results.